sdb/commands/array.py

```diff
@@ -0,0 +1,133 @@
+#
+# Copyright 2020 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Iterable
+
+import drgn
+import sdb
+
+
+class Array(sdb.Command):
+    """
+    Walk through the objects of an array
+
+    EXAMPLES
+        Walk through all the `kmem_cache_node`s of a `kmem_cache`:
+
+            sdb> slabs | head 1 | member node | array
+            *(struct kmem_cache_node *)0xffff90cbd7fba4c0 = {
+                ... <cropped> ...
+            }
+            *(struct kmem_cache_node *)0xffff90cbd7fba4c0 = {
+                ... <cropped> ...
+            }
+            ... <cropped 1022 other elements> ...
+
+        Ask for a specific size when it comes to zero-length arrays:
+
+            sdb> zfs_dbgmsg | head 1 | member zdm_size
+            (int)98
+            sdb> zfs_dbgmsg | head 1 | member zdm_msg
+            (char [1])"s"
+            sdb> zfs_dbgmsg | head 1 | member zdm_msg | array 98
+            warning: exceeding array size 1 - requested 98
+            (char)115
+            (char)112
+            (char)97
+            ... <cropped 95 other elements> ...
+
+        The command also works with pointer arrays as long as the
+        number of elements is specified:
+
+            sdb> spa | member spa_zio_taskq[0][0].stqs_taskq
+            *(taskq_t **)0xffff90cc154dcdc8 = 0xffff90cc127ea600
+
+            sdb> spa | member spa_zio_taskq[0][0].stqs_taskq | array 2
+            *(taskq_t *)0xffff90cc127ea600 = {
+                ... <cropped> ...
+            }
+            *(taskq_t *)0xffff90cc181bfb00 = {
+                ... <cropped> ...
+            }
+
+        Until a better way/notation is implemented on SDB the only
```

Comment:
Out of curiosity, do we need infrastructure work to do this, or is it a matter of adding parsing to the array command to handle something like `array 3-5`?

Reply:
I don't think we need to do anything in the infrastructure. We should be able to bypass argparse (e.g. just use it to give us the leftover tokens that it doesn't understand) and use our own parsing logic.